README.rst
klempner
========
Constructs URLs that targeting other services.

|Version| |Python| |Source| |Coverage| |Quality| |Docs| |CI|

This library makes building URLs for inter-service communication safer
and easier to build.

URL building
------------

.. code-block:: python

   url = klempner.url.build_url('account', 'path', 'with spaces',
                                query='arg',  multi=['arg', 'support'])
   print(url)
   # http://account/path/with%20spaces?query=arg&multi=arg&multi=support

``build_url`` takes care of formatting the path and query parameters correctly
in addition to discovering the service name.  In this example, the service name
is used as-is (see *Unconfigured usage* below).  The real power in ``build_url``
is its ability to discover the scheme, host name, and port number based on the
operating environment.

``build_url`` uses the ``http`` scheme by default.  If the port is determined
by the discovery mechanism, then the scheme is set using a simple global
mapping from port number to scheme.

Discovery examples
------------------

Unconfigured usage
~~~~~~~~~~~~~~~~~~

.. code-block:: python

   url = klempner.url.build_url('account')
   print(url)  # http://account/

This isn't very useful but if you do not configure the discovery mechanism,
then ``build_url`` assumes that the requested service is accessible directly
by name.

Consul service discovery
~~~~~~~~~~~~~~~~~~~~~~~~
The basic form of using consul is not *discovery* at all.  It is simply
URL construction that follows the naming convention that Consul's DNS
interface exposes.

.. code-block:: python

   os.environ['KLEMPNER_DISCOVERY'] = 'consul'
   os.environ['CONSUL_DATACENTER'] = 'production'
   url = klempner.url.build_url('account')
   print(url)  # http://account.service.production.consul/

If you append ``+agent`` to the discovery method, then ``build_url`` will
connect to a Consul agent and retrieve the port number for services.  If the
port has a registered service associated with it, then the service name will
be used as the scheme.

Assuming that the *account* service is registered in consul with a service port
of 8000:

.. code-block:: python

   os.environ['KLEMPNER_DISCOVERY'] = 'consul+agent'
   url = klempner.url.build_url('account')
   print(url)  # http://account.service.production.consul:8000/

Now let's look at what happens for a RabbitMQ connection:

.. code-block:: python

   url = klempner.url.build_url('rabbit')
   print(url)  # amqp://rabbit.service.production.consul:5432/

The scheme is derived by looking up the port in the
``klempner.config.URL_SCHEME_MAP`` and using the result if the lookup
succeeds.

The library will connect to the agent specified by the ``CONSUL_HTTP_ADDR``
environment variable.  If the environment variable is not specified, then the
agent listening on the localhost will be used.

Kubernetes service discovery
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: python

   os.environ['KLEMPNER_DISCOVERY'] = 'kubernetes'
   url = klempner.url.build_url('account')
   print(url)  # http://account.default.svc.cluster.local/

.. code-block:: python

   os.environ['KLEMPNER_DISCOVERY'] = 'kubernetes'
   os.environ['KUBERNETES_NAMESPACE'] = 'my-team'
   url = klempner.url.build_url('account')
   print(url)  # http://account.my-team.svc.cluster.local/

Docker-compose service discovery
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: python

   os.environ['KLEMPNER_DISCOVERY'] = 'docker-compose'
   os.environ['COMPOSE_PROJECT_NAME'] = 'foo'
   url = klempner.url.build_url('account')
   print(url)  # http://127.0.0.1:32867/

This discovery mechanism discovers IP and port numbers for services using
the Docker API.  ``build_url`` retrieves the list of services from the docker
host, filters the list using the "com.docker.compose.project" label, and
selects the service using the "com.docker.compose.service" label.

Environment variable discovery
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
This form of discovery uses environment variables with the service name encoded
into them:

.. code-block:: python

   os.environ['KLEMPNER_DISCOVERY'] = 'environment'
   os.environ['ACCOUNT_HOST'] = '10.2.12.23'
   os.environ['ACCOUNT_PORT'] = '11223'
   url = klempner.url.build_url('account')
   print(url)  # http://10.2.12.23:11223/

For a service named ``adder``, the following environment variables are used
if they are set.

+------------------+-------------------------------+-------------+
| Name             | URL component                 | Default     |
+------------------+-------------------------------+-------------+
| ``ADDER_HOST``   | host portion of the authority | *none*      |
+------------------+-------------------------------+-------------+
| ``ADDER_PORT``   | port portion of the authority | *omitted*   |
+------------------+-------------------------------+-------------+
| ``ADDER_SCHEME`` | scheme                        | *see below* |
+------------------+-------------------------------+-------------+

The URL scheme defaults to looking up the port number in the
``klempner.config.URL_SCHEME_MAP`` dictionary.  If the port number is not
in the dictionary, then ``http`` is used as a default.

.. code-block:: python

   os.environ['KLEMPNER_DISCOVERY'] = 'environment'
   os.environ['ACCOUNT_HOST'] = '10.2.12.23'
   os.environ['ACCOUNT_PORT'] = '443'
   url = klempner.url.build_url('account')
   print(url)  # https://10.2.12.23:443/

Want to contribute?
-------------------
*Thank you.*  See `docs/contributing.rst`_ or
https://klempner.readthedocs.io/en/latest/contributing.html for what you need
to do.

.. _docs/contributing.rst: https://github.com/dave-shawley/klempner/blob
   /master/docs/contributing.rst

.. |CI| image:: https://img.shields.io/circleci/project/github/dave-shawley/klempner/master.svg
   :target: https://circleci.com/gh/dave-shawley/klempner
.. |Coverage| image:: https://img.shields.io/coveralls/github/dave-shawley/klempner.svg
   :target: https://coveralls.io/github/dave-shawley/klempner
.. |Docs| image:: https://img.shields.io/readthedocs/klempner.svg
   :target: https://klempner.readthedocs.io/
.. |Python| image:: https://img.shields.io/pypi/pyversions/klempner.svg
   :target: https://pypi.org/project/klempner
.. |Quality| image:: https://sonarcloud.io/api/project_badges/measure?project=dave-shawley_klempner&metric=alert_status
   :target: https://sonarcloud.io/dashboard?id=dave-shawley_klempner
.. |Source| image:: https://img.shields.io/github/stars/dave-shawley/klempner.svg?logo=github
   :target: https://github.com/dave-shawley/klempner
.. |Version| image:: https://img.shields.io/pypi/v/klempner.svg
   :target: https://pypi.org/project/klempner


docs/history.rst
Release history
===============

Next Release
------------
- :compare:`0.0.3...master`
- Add support for using https with consul by replacing ``CONSUL_HTTP_ADDR``
  with :envvar:`CONSUL_AGENT_URL`

0.0.3 (25 May 2019)
-------------------
- :compare:`0.0.2...0.0.3`
- Add support for the :envvar:`CONSUL_HTTP_TOKEN` environment variable.
- Set ``User-Agent`` when sending requests to Consul.
- Rework documentation.

0.0.2 (19 May 2019)
-------------------
- :compare:`0.0.1...0.0.2`
- Add :data:`~klempner.config.URL_SCHEME_MAP` so that URL schemes can be
  easily configured.
- Separate the :ref:`environment-discovery-method` method out from the
  :ref:`simple-discovery-method` method.  Previously the simple method would
  fall back to using environment variables if they were set.

0.0.1 (5 May 2019)
------------------
- :compare:`0.0.0...0.0.1`
- Implement :ref:`simple-discovery-method`, :ref:`consul-discovery-method`,
  :ref:`consul-agent-discovery-method`, :ref:`kubernetes-discovery-method`
  discovery methods.


tests/helpers.py
import os
import unittest
try:
    import unittest.mock as mock
except ImportError:
    import mock


class EnvironmentMixin(unittest.TestCase):
    """Mix this in to safely manipulate environment variables.

    This mix-in provides methods to manipulate environment variables
    within the test case and restore the original environment in
    ``tearDown``.

    """

    def setUp(self):
        super(EnvironmentMixin, self).setUp()
        self.__environment = {}

    def tearDown(self):
        super(EnvironmentMixin, self).tearDown()
        for name, value in self.__environment.items():
            if value is None:
                os.environ.pop(name, None)
            else:
                os.environ[name] = value

    def setenv(self, name, value):
        """Set the environment variable named `name` to `value`."""
        self.__environment.setdefault(name, os.environ.get(name, None))
        os.environ[name] = value

    def unsetenv(self, name):
        """Clear the environment variable named `name`."""
        self.__environment.setdefault(name, os.environ.get(name, None))
        os.environ.pop(name, None)


class Interceptor(mock.Mock):
    """Mock that intercepts a specific method call and tracks the result.

    This is useful when you want to capture the return value from an
    internal function.  The initial use case was to capture the request
    that was prepared inside of requests.  It turns out that this is
    the easiest way to see exactly which headers are sent short of
    creating a subclass of requests.Session for this purpose.

    This class overwrites `obj.method_name` with itself and uses the
    ``_interceptor`` method as a side effect.  Inside of ``_interceptor``
    it captures the result of calling the original method in ``self.result``
    and returns it.  The outcome is invisible to the inner workings of
    ``obj``, you can inspect the call using the standard mock methods,
    and interrogate the intermediate result as ``self.result``.

    """

    def __init__(self, obj, method_name, *args, **kwargs):
        kwargs['side_effect'] = self._interceptor
        super(Interceptor, self).__init__(*args, **kwargs)
        self.method_name = method_name
        self.obj = obj
        self.result = None
        setattr(obj, method_name, self)

    def _interceptor(self, *args, **kwargs):
        real_method = getattr(self.obj.__class__, self.method_name)
        self.result = real_method(self.obj, *args, **kwargs)
        return self.result

    def _get_child_mock(self, **kwargs):
        # Needed to override this since we do not want to intercept
        # anything for child mocks
        if self._mock_sealed:
            attribute = "." + kwargs["name"] if "name" in kwargs else "()"
            mock_name = self._extract_mock_name() + attribute
            raise AttributeError(mock_name)
        return mock.Mock(**kwargs)


tests/__init__.py
import logging

# swallow log output
logging.getLogger('klempner').addHandler(logging.NullHandler())


tests/test_config.py
import logging
import os
import unittest
try:
    import unittest.mock as mock
except ImportError:
    import mock

from klempner import config, errors, url, version

import tests.helpers


class DirectConfigTests(unittest.TestCase):
    def tearDown(self):
        super(DirectConfigTests, self).tearDown()
        config.reset()

    def test_that_unknown_discovery_method_raises_configuration_error(self):
        with self.assertRaises(errors.ConfigurationError) as context:
            config.configure('unknown-option')
        self.assertEqual('discovery_style',
                         context.exception.configuration_name)
        self.assertEqual('unknown-option',
                         context.exception.configuration_value)

    def test_that_simple_discovery_is_parameterless(self):
        config.configure(config.DiscoveryMethod.SIMPLE)
        _, parameters = config.get_discovery_details()
        self.assertEqual({}, parameters)

    def test_that_consul_discovery_without_datacenter_fails(self):
        with self.assertRaises(errors.ConfigurationError) as context:
            config.configure(config.DiscoveryMethod.CONSUL)
        self.assertEqual('datacenter', context.exception.configuration_name)
        self.assertIs(None, context.exception.configuration_value)

    def test_that_consul_agent_discovery_without_datacenter_fails(self):
        with self.assertRaises(errors.ConfigurationError) as context:
            config.configure(config.DiscoveryMethod.CONSUL_AGENT)
        self.assertEqual('datacenter', context.exception.configuration_name)
        self.assertIs(None, context.exception.configuration_value)


class ConfigByEnvironTests(tests.helpers.EnvironmentMixin, unittest.TestCase):
    def tearDown(self):
        super(ConfigByEnvironTests, self).tearDown()
        config.reset()

    def test_that_unknown_discovery_method_raises_configuration_error(self):
        self.setenv('KLEMPNER_DISCOVERY', 'unknown-option')
        with self.assertRaises(errors.ConfigurationError) as context:
            config.configure_from_environment()
        self.assertEqual('discovery_style',
                         context.exception.configuration_name)
        self.assertEqual('unknown-option',
                         context.exception.configuration_value)

    def test_that_consul_discovery_without_datacenter_fails(self):
        self.setenv('KLEMPNER_DISCOVERY', config.DiscoveryMethod.CONSUL)
        self.unsetenv('CONSUL_DATACENTER')
        with self.assertRaises(errors.ConfigurationError) as context:
            config.configure_from_environment()
        self.assertEqual('CONSUL_DATACENTER',
                         context.exception.configuration_name)
        self.assertIs(None, context.exception.configuration_value)

    def test_that_consul_agent_discovery_without_http_addr_fails(self):
        self.setenv('KLEMPNER_DISCOVERY', config.DiscoveryMethod.CONSUL_AGENT)
        self.unsetenv('CONSUL_HTTP_ADDR')
        with self.assertRaises(errors.ConfigurationError) as context:
            config.configure_from_environment()
        self.assertEqual('CONSUL_HTTP_ADDR',
                         context.exception.configuration_name)
        self.assertIs(None, context.exception.configuration_value)

    def test_that_consul_agent_discovery_uses_consul_token(self):
        self.setenv('KLEMPNER_DISCOVERY', config.DiscoveryMethod.CONSUL_AGENT)
        self.setenv('CONSUL_HTTP_ADDR', '127.0.0.1:1')
        self.setenv('CONSUL_HTTP_TOKEN', 'some-token')
        with mock.patch('klempner.config.requests.get') as requests_get:
            response = mock.Mock()
            response.json.return_value = {'Config': {'Datacenter': 'dc1'}}
            requests_get.return_value = response
            config.configure_from_environment()

        self.assertEqual(1, requests_get.call_count)
        positional, kwargs = requests_get.call_args_list[0]
        self.assertIn('headers', kwargs)
        self.assertIn('Authorization', kwargs['headers'])
        self.assertEqual('bearer some-token',
                         kwargs['headers']['Authorization'].lower())

    @unittest.skipUnless('CONSUL_HTTP_ADDR' in os.environ,
                         'Consul agent is not present')
    def test_that_consul_agent_discovery_includes_user_agent(self):
        self.setenv('KLEMPNER_DISCOVERY', config.DiscoveryMethod.CONSUL_AGENT)
        with mock.patch('klempner.config.requests.get') as requests_get:
            response = mock.Mock()
            response.json.return_value = {'Config': {'Datacenter': 'dc1'}}
            requests_get.return_value = response
            config.configure_from_environment()

        self.assertEqual(1, requests_get.call_count)
        positional, kwargs = requests_get.call_args_list[0]
        self.assertIn('headers', kwargs)
        self.assertIn('User-Agent', kwargs['headers'])
        self.assertEqual('klempner/{}'.format(version),
                         kwargs['headers']['User-Agent'])


class ErrorHandlingTests(unittest.TestCase):
    def test_that_warning_is_logged_for_unused_parameters(self):
        log_messages = []

        class Recorder(logging.Handler):
            def emit(self, record):
                log_messages.append(record.getMessage())

        logger = logging.getLogger('klempner.configure')
        handler = Recorder()
        logger.addHandler(handler)
        self.addCleanup(logger.removeHandler, handler)

        config.configure(config.DiscoveryMethod.SIMPLE, unusued='parameter')
        for message in log_messages:
            if 'simple does not accept additional parameters' in message:
                break
        else:
            self.fail('expected to find warning message in ' +
                      repr(log_messages))


class SchemeMappingTests(tests.helpers.EnvironmentMixin, unittest.TestCase):
    def setUp(self):
        self._saved_scheme_map = config.URL_SCHEME_MAP.copy()
        super(SchemeMappingTests, self).setUp()

    def tearDown(self):
        super(SchemeMappingTests, self).tearDown()
        config.URL_SCHEME_MAP.clear()
        config.URL_SCHEME_MAP.update(self._saved_scheme_map)

    def test_that_mapping_can_be_disabled(self):
        config.URL_SCHEME_MAP.clear()
        config.configure(config.DiscoveryMethod.ENV_VARS)
        self.setenv('ACCOUNT_HOST', 'account.example.com')
        self.setenv('ACCOUNT_PORT', '443')
        self.assertEqual('http://account.example.com:443/',
                         url.build_url('account'))

    def test_that_mapping_can_be_overridden(self):
        config.URL_SCHEME_MAP[5672] = 'rabbitmq'
        config.configure(config.DiscoveryMethod.ENV_VARS)
        self.setenv('ACCOUNT_HOST', 'account.example.com')
        self.setenv('ACCOUNT_PORT', '5672')
        self.assertEqual('rabbitmq://account.example.com:5672/',
                         url.build_url('account'))


klempner/url.py
from __future__ import unicode_literals

import logging
import os

import requests.adapters

import cachetools
from klempner import compat, config, errors, version

#    pchar         = unreserved / pct-encoded / sub-delims / ":" / "@"
#    sub-delims    = "!" / "$" / "&" / "'" / "(" / ")"
#                  / "*" / "+" / "," / ";" / "="
#    unreserved    = ALPHA / DIGIT / "-" / "." / "_" / "~"

PATH_SAFE_CHARS = ":@!$&'()*+,;=-._~"
"""Safe characters for path elements."""


class State(object):
    """Module state.

    A single instance of this class exists as a module-level property.
    It caches information as it is discovered.  Applications SHOULD
    call :func:`.reset_cache` if they suspect that the discovery
    configuration has changed.

    """

    def __init__(self):
        self.discovery_cache = cachetools.TTLCache(50, 300)
        self.logger = logging.getLogger(__package__)
        self.session = self._create_session()

    def clear(self):
        self.discovery_cache.clear()
        self.session.close()
        self.session = self._create_session()

    def lookup_consul_service(self, service):
        sentinel = object()
        service_info = self.discovery_cache.get(service, sentinel)
        if service_info is sentinel:
            parsed = compat.urlparse(os.environ['CONSUL_AGENT_URL'])
            url = compat.urlunparse(
                (parsed[0], parsed[1],
                 '/v1/catalog/service/{0}'.format(service), '', '', ''))
            headers = {}
            if os.environ.get('CONSUL_HTTP_TOKEN'):
                headers['Authorization'] = 'Bearer {0}'.format(
                    os.environ['CONSUL_HTTP_TOKEN'])

            response = self.session.get(url, headers=headers)
            response.raise_for_status()
            body = response.json()
            if body:
                service_info = body[0]
                self.discovery_cache[service] = service_info
            else:
                service_info = None

        return service_info

    @staticmethod
    def _create_session():
        session = requests.Session()
        session.headers['User-Agent'] = '/'.join([__package__, version])
        return session


_state = State()


def reset_cache():
    """Reset internal caches.

    Applications MUST call this function if they have changed discovery
    configuration details or suspect that they may have changed.  This
    should not happen often since the discovery configuration is based
    primarily on environment variables which are not modifiable from
    outside of the process.

    """
    config.reset()
    _state.clear()


def build_url(service, *path, **query):
    """Build a URL that targets `service`.

    :param str service: service to target
    :param path: request path elements
    :param query: request query parameters
    :returns: a fully-formed, absolute URL
    :rtype: str

    """
    config.ensure_configured()
    buf = compat.StringIO()
    _write_network_portion(buf, service)
    buf.write('/')
    buf.write('/'.join(
        compat.quote(str(p), safe=PATH_SAFE_CHARS) for p in path))

    query_tuples = []
    for name, value in query.items():
        if isinstance(value, compat.Mapping):
            raise ValueError('Mapping query parameters are unsupported')
        if (isinstance(value, compat.Iterable)
                and not isinstance(value, compat.TEXT_TYPES)):
            query_tuples.extend((name, elm) for elm in sorted(value))
        else:
            query_tuples.append((name, value))
    if query_tuples:
        query_tuples.sort()
        buf.write('?')
        buf.write('&'.join(
            '{0}={1}'.format(_quote_query_arg(name), _quote_query_arg(value))
            for name, value in query_tuples))

    return buf.getvalue()


def _write_network_portion(buf, service):
    """Add the discovered network portion to `buf`.

    :param klempner.compat.StringIO buf: buffer to write the discovered
        network details to
    :param str service: name of the service that is being looked up

    """
    env_service = service.upper()
    discovery_style, parameters = config.get_discovery_details()
    if discovery_style == config.DiscoveryMethod.CONSUL:
        buf.write('http://')
        buf.write(service)
        buf.write('.service.')
        buf.write(parameters['datacenter'])
        buf.write('.consul')
    elif discovery_style == config.DiscoveryMethod.CONSUL_AGENT:
        service_info = _state.lookup_consul_service(service)
        if not service_info:  # service does not exist in consul
            raise errors.ServiceNotFoundError(service)
        else:
            calculated_scheme = config.URL_SCHEME_MAP.get(
                service_info['ServicePort'], 'http')
            meta = service_info.get('ServiceMeta', {})
            buf.write(meta.get('protocol', calculated_scheme))
            buf.write('://')
            buf.write(service_info['ServiceName'])
            buf.write('.service.')
            buf.write(service_info['Datacenter'])
            buf.write('.consul:')
            buf.write(str(service_info['ServicePort']))
    elif discovery_style == config.DiscoveryMethod.K8S:
        buf.write('http://')
        buf.write(service + '.')
        buf.write(parameters['namespace'])
        buf.write('.svc.cluster.local')
    elif discovery_style == config.DiscoveryMethod.ENV_VARS:
        scheme = os.environ.get('{0}_SCHEME'.format(env_service), None)
        host = os.environ.get('{0}_HOST'.format(env_service), None)
        port = os.environ.get('{0}_PORT'.format(env_service), None)

        if port is not None and port.startswith('tcp://'):
            # special case for docker's ip:port format
            parts = compat.urlparse(port)
            port = str(parts.port)
            if host is None:
                host = parts.hostname
        if scheme is None:
            if port is not None:
                scheme = config.URL_SCHEME_MAP.get(int(port), 'http')
            else:
                scheme = 'http'
        buf.write(scheme)
        buf.write('://')
        buf.write(host or service)
        if port is not None:
            buf.write(':')
            buf.write(port)
    else:
        buf.write('http://')
        buf.write(service)


def _quote_query_arg(v):
    if not isinstance(v, compat.TEXT_TYPES):
        v = str(v)
    return compat.quote(v.encode('utf-8'))


klempner/config.py
import logging
import os

import requests

from klempner import compat, errors, version

URL_SCHEME_MAP = {
    5672: 'amqp',  # https://www.rabbitmq.com/uri-spec.html
    21: 'ftp',  # https://tools.ietf.org/html/rfc1738
    70: 'gopher',  # https://tools.ietf.org/html/rfc4266
    80: 'http',  # https://tools.ietf.org/html/rfc7230#section-2.7.1
    443: 'https',  # https://tools.ietf.org/html/rfc7230#section-2.7.2
    1344: 'icap',  # https://tools.ietf.org/html/rfc3507#section-4.2
    631: 'ipp',  # https://tools.ietf.org/html/rfc3510
    389: 'ldap',  # https://tools.ietf.org/html/rfc4516
    636: 'ldaps',  # https://tools.ietf.org/html/rfc4516
    # https://docs.mongodb.com/manual/reference/connection-string/
    27017: 'mongodb',
    3306: 'mysql',
    119: 'nntp',  # https://tools.ietf.org/html/rfc5538
    110: 'pop',  # https://tools.ietf.org/html/rfc2384
    # https://www.postgresql.org/docs/current/libpq-connect.html#LIBPQ-CONNSTRING
    5432: 'postgresql',
    6379: 'redis',  # https://www.iana.org/assignments/uri-schemes/prov/redis
    873: 'rsync',  # https://tools.ietf.org/html/rfc5781
    554: 'rtsp',  # https://tools.ietf.org/html/rfc7826#section-4.2
    322: 'rtsps',  # https://tools.ietf.org/html/rfc7826#section-4.2
    25: 'smtp',  # https://tools.ietf.org/html/draft-melnikov-smime-msa-to-mda
    161: 'snmp',  # https://tools.ietf.org/html/rfc4088
    22: 'ssh',  # https://tools.ietf.org/html/draft-ietf-secsh-scp-sftp-ssh-uri
    23: 'telnet',  # https://tools.ietf.org/html/rfc4248
    69: 'tftp',  # https://tools.ietf.org/html/rfc3617
    3372: 'tip',  # https://tools.ietf.org/html/rfc2371
    5900: 'vnc',  # https://tools.ietf.org/html/rfc7869
    602: 'xmlrpc.beep',  # https://tools.ietf.org/html/rfc3529#section-5.1
}
"""Mapping of port number to URL scheme.

This dictionary is used to identify the URL scheme based on the port number
when a port number is available.  Users of the library MAY modify the content
of this dictionary **at any time**.

"""


class DiscoveryMethod(object):
    """Available discovery methods."""

    SIMPLE = 'simple'
    """Build URLs using service name as host name."""

    CONSUL = 'consul'
    """Build consul-based service URLs."""

    CONSUL_AGENT = 'consul+agent'
    """Build consul-based service URLs using a consul agent."""

    ENV_VARS = 'environment'
    """Build URLs based on _HOST, _PORT, and _SCHEME environment variables."""

    K8S = 'kubernetes'
    """Build Kubernetes cluster-based service URLs."""

    DEFAULT = SIMPLE

    UNSET = object()

    AVAILABLE = (CONSUL, CONSUL_AGENT, ENV_VARS, K8S, SIMPLE, UNSET)


_discovery_method = DiscoveryMethod.UNSET
_discovery_parameters = {}


def reset():
    """Reset URL construction parameters."""
    configure(DiscoveryMethod.UNSET)


def ensure_configured():
    """Configure from the environment if currently unconfigured."""
    if _discovery_method is DiscoveryMethod.UNSET:
        configure_from_environment()


def configure(discovery_method, **parameters):
    """Configure the discovery method directly.

    :param discovery_method: method to use
    :param parameters: parameters required for the selected
        method
    :raises: :exc:`klempner.errors.ConfigurationError` if a required
        parameter is not provided

    """
    logger = logging.getLogger(__package__).getChild('configure')

    def require_parameter(name):
        try:
            return parameters.pop(name)
        except KeyError:
            logger.error('parameter %s is required by discovery method %s',
                         name, discovery_method)
            raise errors.ConfigurationError(name, None)

    logger.debug('configuring for discovery_method %s with parameters=%r',
                 discovery_method, parameters)
    incoming_parameters = {}
    if discovery_method == DiscoveryMethod.CONSUL:
        incoming_parameters['datacenter'] = require_parameter('datacenter')
    elif discovery_method == DiscoveryMethod.CONSUL_AGENT:
        incoming_parameters['datacenter'] = require_parameter('datacenter')
    elif discovery_method == DiscoveryMethod.K8S:
        incoming_parameters['namespace'] = require_parameter('namespace')
    elif discovery_method not in DiscoveryMethod.AVAILABLE:
        raise errors.ConfigurationError('discovery_style', discovery_method)

    global _discovery_method, _discovery_parameters
    if _discovery_method is DiscoveryMethod.UNSET:
        logger.info('setting discovery method to %r with parameters=%r',
                    discovery_method, incoming_parameters)
    else:
        logger.info(
            'setting discovery method: current_method=%r new_method=%r '
            'parameters=%r', _discovery_method, discovery_method,
            incoming_parameters)
    _discovery_method = discovery_method
    _discovery_parameters.clear()
    _discovery_parameters.update(incoming_parameters)
    if parameters:
        logger.warning(
            'discovery style %s does not accept additional parameters, '
            '%d extra parameters were passed to configure', discovery_method,
            len(parameters))


def configure_from_environment():
    """Set the discovery method from ``$KLEMPNER_DISCOVERY``.

    This method configures the library based on the
    :envvar:`KLEMPNER_DISCOVERY` environment variable.  If the
    environment variable is not set, then the "simple" configuration
    is used.

    """
    logger = logging.getLogger(__package__).getChild(
        'configure_from_environment')
    new_method = os.environ.get('KLEMPNER_DISCOVERY', DiscoveryMethod.SIMPLE)
    logger.debug('configuring from environment: discovery_method=%s',
                 new_method)

    def require_envvar(name):
        try:
            return os.environ[name]
        except KeyError:
            logger.error(
                'discovery method %s requires the %s environment variable',
                new_method, name)
            raise errors.ConfigurationError(name, None)

    parameters = {}
    if new_method == DiscoveryMethod.CONSUL:
        parameters['datacenter'] = require_envvar('CONSUL_DATACENTER')
    elif new_method == DiscoveryMethod.CONSUL_AGENT:
        headers = {'User-Agent': '/'.join([__package__, version])}
        try:
            headers['Authorization'] = 'Bearer {}'.format(
                os.environ['CONSUL_HTTP_TOKEN'])
        except KeyError:
            pass
        parsed = compat.urlparse(require_envvar('CONSUL_AGENT_URL'))
        url = compat.urlunparse((parsed[0], parsed[1], '/v1/agent/self', None,
                                 None, None))
        response = requests.get(url, headers=headers)
        response.raise_for_status()
        body = response.json()
        parameters['datacenter'] = body['Config']['Datacenter']
    elif new_method == DiscoveryMethod.K8S:
        parameters['namespace'] = os.environ.get('KUBERNETES_NAMESPACE',
                                                 'default')
    elif new_method not in DiscoveryMethod.AVAILABLE:
        raise errors.ConfigurationError('discovery_style', new_method)

    configure(new_method, **parameters)


def get_discovery_details():
    """Retrieve the configured method and parameters.

    :rtype: tuple(str, dict)

    """
    return _discovery_method, _discovery_parameters.copy()
